File: maximum-number-of-vowels-in-a-substring-of-given-length/maximum-number-of-vowels-in-a-substring-of-given-length.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var maxVowels = function(s, k) {
    var numVowels = 0
    var vowels = 'aeiou'

    for (let i=0; i<k; i++) {
        char = s[i]
        if (vowels.includes(char)) {
            numVowels += 1
        }
    }

    // check if input string equals k, means we already iterated through entire string
    // or if found the max number of vowels possible, which is k itself
    if ((numVowels === k) || (s.length === k)) {
        return numVowels
    }

    var maxVowels = numVowels

    for (let i=k; i<s.length; i++) {
        var newChar = s[i]
        var oldChar = s[i-k]
        numVowels += vowels.includes(oldChar) ? -1 : 0
        numVowels += vowels.includes(newChar) ? 1 : 0

        if (numVowels > maxVowels) {
            maxVowels = numVowels
        }

        if (numVowels === k) {
            return maxVowels
        }
    }
    
    return maxVowels
};

File: maximum-number-of-vowels-in-a-substring-of-given-length/maximum-number-of-vowels-in-a-substring-of-given-length.py
class Solution(object):
    def maxVowels(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: int
        """
        vowels = 'aeiou'
        numVowels = 0

        # count number of vowels in initial substring
        for char in s[:k]:
            if char in vowels:
                numVowels += 1
            
        # return number of vowels if we already tested the entire string 
        # or if we've reached max num of vowels
        if (numVowels == k) or (len(s) == k):
            return numVowels

        maxVowels = numVowels

        # sliding window of substring
        for i in range(k, len(s)):
            newChar = s[i]
            oldChar = s[i-k]
            numVowels += 1 if newChar in vowels else 0
            numVowels += -1 if oldChar in vowels else 0     

            if numVowels > maxVowels:
                maxVowels = numVowels

            if numVowels == k:
                return numVowels

        return maxVowels
